packages/flutter/lib/src/widgets/text_editing_action.dart

```diff
@@ -0,0 +1,64 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import 'actions.dart';
+import 'editable_text.dart';
+import 'focus_manager.dart';
+import 'framework.dart';
+
+/// Similar to [CallbackAction]'s [OnInvokeCallback], but includes
+/// [EditableTextState] as a parameter.
+typedef OnInvokeTextEditingCallback<T extends Intent> = Object? Function(T intent, EditableTextState editableTextState);
+
+/// An [Action] related to editing text.
+///
+/// If an [EditableText] is currently focused, then it will be passed to the
+/// given [onInvoke] callback. If not, then [isEnabled] will be false and
+/// [onInvoke] will not be called.
+///
+/// See also:
+///
+///  * [CallbackAction], which is a similar Action type but unrelated to text
+///    editing.
+class TextEditingAction<T extends Intent> extends ContextAction<T> {
+  /// A constructor for a [TextEditingAction].
+  ///
+  /// The [onInvoke] parameter must not be null.
+  TextEditingAction({ required this.onInvoke }) : assert(onInvoke != null);
+
+  EditableTextState? get _editableTextState {
+    // If an EditableText is not focused, then ignore this action.
+    if (primaryFocus?.context?.widget is! EditableText) {
```

Comment:
This is tying TextEditingAction tightly to EditableText. If somebody would want to implement their own EditableText, they'd have to re-implement all actions as well. Could we decouple this by defining an TextEditingActionExecutor interface (maybe better name needed) that defines the methods a TextEditingAction can operate on? EditableTextState would then implement that interface.

Reply:
That's a good catch, thanks.  I called it TextEditingActionTarget and tried to do it like this: fb6970b

Reply:
looks like this still needs to be replaced with a proper type check to match the down cast below?